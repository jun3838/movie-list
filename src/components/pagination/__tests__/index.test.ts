import { mount } from '@vue/test-utils'
import { nextTick } from 'vue'
import Pagination from '../index.vue'

describe('should render pagination correctly', () => {
  test('with less than maxVisibleButton', () => {
    const wrapper = mount(Pagination, {
      props: {
        totalPages: 4,
        currentPage: 1
      }
    })

    expect(wrapper.html()).toMatchSnapshot()
  })

  test('with more than maxVisibleButton', () => {
    const wrapper = mount(Pagination, {
      props: {
        totalPages: 10,
        currentPage: 1
      }
    })

    expect(wrapper.html()).toMatchSnapshot()
  })
})

describe('computed', () => {
  test('startPage', async () => {
    const wrapper = mount(Pagination, {
      props: {
        totalPages: 10,
        currentPage: 1
      }
    })

    expect(wrapper.vm.startPage).toEqual(1)
    wrapper.setProps({ currentPage: 5})
    await nextTick()
    expect(wrapper.vm.startPage).toEqual(3)
  })
  test('endPage', async () => {
    const wrapper = mount(Pagination, {
      props: {
        totalPages: 10,
        currentPage: 1
      }
    })

    expect(wrapper.vm.endPage).toEqual(5)
    wrapper.setProps({ currentPage: 3})
    await nextTick()
    expect(wrapper.vm.endPage).toEqual(5)
  })
  test('pages', async () => {
    const wrapper = mount(Pagination, {
      props: {
        totalPages: 10,
        currentPage: 1
      }
    })

    expect(wrapper.vm.pages).toEqual([1, 2, 3, 4, 5])
    wrapper.setProps({ currentPage: 8})
    await nextTick()
    expect(wrapper.vm.pages).toEqual([6, 7, 8, 9, 10])
  })
  test('isFirstPage', async () => {
    const wrapper = mount(Pagination, {
      props: {
        totalPages: 10,
        currentPage: 1
      }
    })

    expect(wrapper.vm.isFirstPage).toBeTruthy()
    wrapper.setProps({ currentPage: 2})
    await nextTick()
    expect(wrapper.vm.isFirstPage).toBeFalsy()
  })
  test('isLastPage', async () => {
    const wrapper = mount(Pagination, {
      props: {
        totalPages: 10,
        currentPage: 1
      }
    })

    expect(wrapper.vm.isLastPage).toBeFalsy()
    wrapper.setProps({ currentPage: 10})
    await nextTick()
    expect(wrapper.vm.isLastPage).toBeTruthy()
  })
})

describe('emit', () => {
  test('should emit "next"', async () => {
    const wrapper = mount(Pagination, {
      props: {
        totalPages: 10,
        currentPage: 1
      }
    })

    wrapper.find('.next').trigger('click')
    expect(wrapper.emitted()).toHaveProperty('next')
  })
  test('should emit "prev"', async () => {
    const wrapper = mount(Pagination, {
      props: {
        totalPages: 10,
        currentPage: 3
      }
    })

    wrapper.find('.prev').trigger('click')
    expect(wrapper.emitted()).toHaveProperty('prev')
  })
  test('should emit "goto"', async () => {
    const wrapper = mount(Pagination, {
      props: {
        totalPages: 10,
        currentPage: 1
      }
    })

    wrapper.findAll('.page')[2].trigger('click')
    expect(wrapper.emitted().goto[0]).toEqual([3])

    wrapper.find('.first').trigger('click')
    expect(wrapper.emitted().goto[1]).toEqual([1])

    wrapper.find('.last').trigger('click')
    expect(wrapper.emitted().goto[2]).toEqual([10])
  })
})
