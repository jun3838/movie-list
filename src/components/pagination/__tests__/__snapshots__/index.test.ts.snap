// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render pagination correctly with less than maxVisibleButton 1`] = `
"<ul class=\\"flex\\">
  <li class=\\"pagination-button first disabled\\"> &lt;&lt; </li>
  <li class=\\"pagination-button prev disabled\\"> &lt; </li>
  <li class=\\"active pagination-button page\\">1</li>
  <li class=\\"pagination-button page\\">2</li>
  <li class=\\"pagination-button page\\">3</li>
  <li class=\\"pagination-button page\\">4</li>
  <li class=\\"pagination-button next\\"> &gt; </li>
  <li class=\\"pagination-button last\\"> &gt;&gt; </li>
</ul>"
`;

exports[`should render pagination correctly with more than maxVisibleButton 1`] = `
"<ul class=\\"flex\\">
  <li class=\\"pagination-button first disabled\\"> &lt;&lt; </li>
  <li class=\\"pagination-button prev disabled\\"> &lt; </li>
  <li class=\\"active pagination-button page\\">1</li>
  <li class=\\"pagination-button page\\">2</li>
  <li class=\\"pagination-button page\\">3</li>
  <li class=\\"pagination-button page\\">4</li>
  <li class=\\"pagination-button page\\">5</li>
  <li class=\\"pagination-button next\\"> &gt; </li>
  <li class=\\"pagination-button last\\"> &gt;&gt; </li>
</ul>"
`;
